mis/pipeline/maximization.py

```diff
@@ -0,0 +1,128 @@
+
+    This postprocessor expects that a result could be vulnerable to bitflips, so it
+    will attempt to fix any result provided by the quantum algorithm, to make it
+    independent (if it's not independent) and maximal (if it's not maximal).
+    """
+
+    def __init__(
+        self,
+        frequency_threshold: float = 1e-7,
+        augment_rounds: int = 10,
+        seed: int = 0,
+    ):
+        """
+        frequency_threshold: Discard any solution which show up with a frequency
+            <= frequency_threshold. Set 0 to never discard any solution.
+        augment_rounds: The number of attempts to augment an independent set to
+            add possibly missing nodes.
+        seed: A random seed.
+        """
+        self.frequency_threshold = frequency_threshold
+        self.augment_rounds = augment_rounds
+        self.seed = seed
+
+    def postprocess(self, solution: MISSolution) -> MISSolution | None:
+        """
+        The main entry point: attempt to improve a solution.
+        """
+        if solution.frequency < self.frequency_threshold:
+            return None
+        if not self.is_independent_solution(solution):
+            solution = self.reduce_to_independence(solution)
+        # From this point, we're sure that `solution` is independent.
+        if nx.is_dominating_set(solution.original, solution.nodes):
+            # Maximum independent set, we can't do better.
+            return solution
+        return self.augment_to_maximal(solution)
+
+    def is_independent_solution(self, solution: MISSolution) -> bool:
+        """
+        Check whether a solution is independent.
+        """
+        return self.is_independent_list(graph=solution.original, nodes=solution.nodes)
+
+    def is_independent_list(self, graph: nx.Graph, nodes: list[int]) -> bool:
+        """
+        Check whether a list of nodes within a graph is independent.
+        """
+        for i, u in enumerate(nodes):
+            for v in nodes[i:]:
+                if graph.has_edge(u, v):
+                    return False
+        return True
+
+    def augment_to_maximal(self, solution: MISSolution) -> MISSolution:
+        """Augment a given set up to a maximal IS using a greedy algorithm running k times.
+
+        See https://doi.org/10.48550/arXiv.2202.09372 section 2.3 of supplementary material for reference.
+        """
+        unpicked_nodes = set(solution.original.nodes) - set(solution.nodes)
+
+        # The best solution so far.
+        best_pick = solution.nodes
+
+        rng = random.Random(self.seed)
+        for _ in range(self.augment_rounds):
+            # Pick an arbitrary and random order.
+            order = list(unpicked_nodes)
+            rng.shuffle(order)
+
+            # Attempt to grow the list of nodes in this order.
+            picked = list(solution.nodes)
+            for node in order:
+                maybe_picked = list(picked)
```

Comment:
Isn't `picked` already a list at this stage ?

Reply:
Yeah, we're making a copy.